### PR TITLE
yggdrasil: Change package configuration to UCI

### PR DESCRIPTION
**Pull request by zhoreeq**
Maintainer: @wfleurant
Compile tested: mips / snapshot

Description:

[WIP] Adding Luci application to configure yggdrasil via web interface

### CODE REVIEW

**Comment by wfleurant** [comment]
Please bump the package release version (in Makefile) to be: `PKG_RELEASE:=4`

**Comment by zhoreeq** [comment]
Done

File: net/yggdrasil/files/yggdrasil.init [hunk 1]
```diff
@@ -12,7 +12,7 @@ start_service()
 	procd_open_instance
 	procd_set_param respawn
 
-	procd_set_param command /usr/sbin/yggdrasil -useconffile /etc/yggdrasil.conf
+	procd_set_param command /bin/ash -c "ygguci get | yggdrasil -useconf"
```
Comment (wfleurant) [inline net/yggdrasil/files/yggdrasil.init]: This seems correct. OpenWrt procd should always restart with latest UCI settings.  

The code below _does not_  re-write the `/etc/config/yggdrasil` file if yggdrasil is killed or crashes. The `/etc/init.d/yggdrasil` function `start_service()`  is not called after a crash or `killall yggdrasil`. Only `procd_set_param` values are re-executed.

```
start_service()
{
  procd_open_instance
  procd_set_param respawn

  ygguci get > /tmp/etc/yggdrasil.conf
  procd_set_param command /usr/sbin/yggdrasil -useconffile /tmp/etc/yggdrasil.conf
....
```

Is there a way we can keep `procd_set_param` to a single command, but always regenerate configs from uci on crash or kill? Do we keep the example code because it re-writes the config (as intended) with `/etc/init.d/yggdrasil restart`?

File: net/yggdrasil/files/ygguci [hunk 2]
```diff
@@ -0,0 +1,247 @@
...
+
+  for interface, peers in pairs(obj.InterfacePeers) do
+      for _, v in pairs(peers) do
+          local name = cursor:add("yggdrasil", "interface_peer")
+          cursor:set("yggdrasil", name, "interface", interface)
+          cursor:set("yggdrasil", name, "uri", v)
+      end
+  end
+
+  cursor:set("yggdrasil", "NodeInfo", "yggdrasil")
```
Comment (wfleurant) [inline net/yggdrasil/files/ygguci]: Can we please lowercase all UCI section names? As far as I see this is the only occurrence for sections.

File: net/yggdrasil/files/ygguci [hunk 3]
```diff
@@ -0,0 +1,247 @@
...
+function UCI.set(obj)
+  local cursor = uci.cursor()
+
+  for i, section in pairs(cursor:get_all("yggdrasil")) do
+    cursor:delete("yggdrasil", section[".name"])
+  end
+
+
+  cursor:set("yggdrasil", "yggdrasil", "yggdrasil")
+  cursor:set("yggdrasil", "yggdrasil", "EncryptionPublicKey", obj.EncryptionPublicKey)
```
Comment (wfleurant) [inline net/yggdrasil/files/ygguci]: Can we please lowercase all UCI options names? All uci options (as far as i can tell) should be lower case.

Comment (zhoreeq) [reply net/yggdrasil/files/ygguci]: Do you think lowercase option names is a better common practice? FYI /etc/config/dropbear also has CamelCase option names. Those settings are direct mappings for the same keys in yggdrasil.conf, so I thought it could make more sense.

File: net/yggdrasil/files/ygguci (hunk 2 above)
Comment (zhoreeq) [reply net/yggdrasil/files/ygguci]: Done

File: net/yggdrasil/files/ygguci (hunk 3 above)
Comment (wfleurant) [reply net/yggdrasil/files/ygguci]: Could flip a coin either way.

File: net/yggdrasil/files/ygguci [hunk 4]
```diff
@@ -81,7 +81,7 @@ function UCI.get()
       table.insert(obj.InterfacePeers[s["interface"]], s.uri)
   end)
 
-  local node_info = cursor:get_all("yggdrasil", "NodeInfo")
```
Comment (wfleurant) [inline net/yggdrasil/files/ygguci]: OK -- If we want to try to match yggdrasil.config to uci let's keep this patch out of the mix.

Comment (wfleurant) [reply net/yggdrasil/files/ygguci]: Just leave NodeInfo the way it is, and we'll try to keep the same case between config and UCI

File: net/yggdrasil/files/yggdrasil.init (hunk 1 above)
Comment (wfleurant) [reply net/yggdrasil/files/yggdrasil.init]: @neheb could you please advise how to handle our procd launching yggdrasil? We had changed procd launching yggdrasil this way because we simply had /etc/yggdrasil.conf. Currently, we're looking for help to launch with the new `/usr/bin/ygguci` handler code.  

Ideas?

File: net/yggdrasil/files/ygguci [hunk 5]
```diff
@@ -0,0 +1,247 @@
+#!/usr/bin/env lua
```
Comment (feckert) [inline net/yggdrasil/files/ygguci]: This comment is out of scope.
But could we please convert <spaces> into <tabs> in this file?
OpenWrt always uses <tabs>.
This reduces the file size.

File: net/yggdrasil/files/yggdrasil.init (hunk 1 above)
Comment (wfleurant) [reply net/yggdrasil/files/yggdrasil.init]: @feckert -- any ideas? I'm thinking this is/was an issue previously but now I don't have a solution going forward?

Comment (feckert) [reply net/yggdrasil/files/yggdrasil.init]: Why do we need to recreate the `/tmp/etc/yggdrasil.conf` file, if the process dies and the service is restarted by `procd`. There are no changes in the uci config?

**Comment by feckert** [comment]
An additional hint on thies pullrequest. We could move the firewall config from yggdrasil.defaults also to procd. So we do not need an uci-default script to add firewall uci rules. Please have a look to https://github.com/openwrt/packages/blob/665ec53cd8bc616ae132b2957dac4f30e072a909/net/net-snmp/files/snmpd.init#L262

**Comment by zhoreeq** [comment]
@feckert maybe we can finish this and create another PR for dealing with firewall stuff?

File: net/yggdrasil/files/yggdrasil.init (hunk 1 above)
Comment (feckert) [reply net/yggdrasil/files/yggdrasil.init]: ping @wfleurant

**Comment by feckert** [comment]
@zhoreeq yes we could do this. I will merge if @wfleurant is answering my question.
https://github.com/openwrt/packages/pull/10547#discussion_r347826563

File: net/yggdrasil/files/yggdrasil.init (hunk 1 above)
Comment (zhoreeq) [reply net/yggdrasil/files/yggdrasil.init]: We can also add -normaliseconf call after ygguci get. 

```
/bin/ash -c "ygguci get | yggdrasil -useconf -normaliseconf -json | yggdrasil -useconf"
```

This way it will always run with correct config even if newer versions of yggdrasil will introduce some radical config changes. What do you think?

Comment (feckert) [reply net/yggdrasil/files/yggdrasil.init]: I am not familar with you tool.
If we need this then i am fine.
My question was why do we need to recreate the /tmp/etc/yggdrasil.conf from uci if we get prespawned by procd?
I thought that was the intention by @wfleurant ?

Comment (wfleurant) [reply net/yggdrasil/files/yggdrasil.init]: No, you're right. There is no need to re/create `/tmp/etc/yggdrasil.conf`  

The reason it is created stems from a feature in luci-app-cjdns. It allows the manual editing of config from LuCI. We're not there yet re: `luci-app-yggdrasil`

Comment (zhoreeq) [reply net/yggdrasil/files/yggdrasil.init]: @wfleurant good to have you back! I'm testing the luci app already, check your DMs in Matrix. Changing settings via LuCI works just fine.

What do you think about -normaliseconf, should I commit this change? I think it could be useful, they do that trick in Debian packages.

Comment (wfleurant) [reply net/yggdrasil/files/yggdrasil.init]: Yes, you're right. It's best to use -normaliseconf -- re: Matrix.
> yup, it sets every setting possible. one thing may break it, if a user will set something stupid manually without a web interface
> The -normaliseconf Yggdrasil option lets you sanitise the config to avoid this kind of problem

**Comment by feckert** [comment]
@zhoreeq Before I could merge this please stash/fixup your commits into one meaningful commits.
so we have only **yggdrasil: add uci handling**

**Comment by zhoreeq** [comment]
@feckert Can it be done in this branch or should I create another one with a single commit?

**Comment by zhoreeq** [comment]
OK, now it's correct

**Comment by feckert** [comment]
@zhoreeq thanks for your contribution
@wfleurant thanks for your review.
I hope we can move the firewall config into procd.

**Comment by Dimazhan** [comment]
hi
after sysupgrade and install yggdrasil I received an error:
```
Configuring yggdrasil.
dkjson: no valid JSON value at line 2, column 3
```
and configuration file /etc/config/yggdrasil was created with a size of 0 bytes
the error was caused by the presence of comments in the configuration file

line 28 in yggdrasil/files/yggdrasil.defaults
`cat /etc/yggdrasil.conf | ygguci set`
needs to be corrected to
`yggdrasil -useconffile /etc/yggdrasil.conf -normaliseconf -json | ygguci set`
